keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -0,0 +1,355 @@
+- [Performance degradation in large k8s deployments.][perf-degrade] Not being able to efficiently read/update individual endpoint changes can lead to (e.g during rolling upgrade of a service) endpoints operations that are quadratic in the number of its elements. If one consider watches in the picture (there's one from each kube-proxy), the situation becomes even worse as the quadratic traffic gets multiplied further with number of watches (usually equal to #nodes in the cluster).
+
+The new EndpointSlice API aims to address existing problems as well as leaving room for future extension.
+
+
+### Goal
+
+- Support tens of thousands of backend endpoints in a single service on cluster with thousands of nodes.
+- Move the API towards a general-purpose backend discovery API.
+- Leave room for foreseeable extension:
+  - Support multiple IPs per pod
+  - More endpoint states than Ready/NotReady
+  - Dynamic endpoint subsetting
+  
+### Non-Goal
+- Change functionality provided by K8s V1 Service API.
+- Provide better load balancing for K8s service backends.
+
+## Proposal
+
+### EndpointSlice API
+The following new EndpointSlice API will be added to the `Discovery` API group.
+
+```
+type EndpointSlice struct {
+        metav1.TypeMeta
+        // OwnerReferences should be set when the object is derived from a service.
+        metav1.ObjectMeta
+
+        Endpoints []Endpoint
+        // Each EndpiontPort must have a unique port name
+        Ports []EndpointPort
+}
+
+type EndpointPort struct {
+        // Required: The name of this port.
+        // Must be a DNS_LABEL or an empty string.
+        Name string 
+        // Required: The IP protocol for this port.
+        // Must be UDP, TCP, or SCTP.
+        // Default is TCP.
+        Protocol v1.Protocol
+        // Optional: The port number of the endpoint.
+        // If this is not specified, ports are not restricted and must be interpreted in the context of the specific consumer.
+        Port *int32
+}
+
+type Endpoint struct {
+        // Required: must contain at least one backend.
+        // This can be an IP, URL or hostname.
+        // Different consumers (e.g. kube-proxy) handle different types of backends in the context of its own capabilities.
+        Backends []string
+        // Optional: The Hostname of this endpoint
+        Hostname string         
+        // Optional: Node hosting this endpoint. This can be used to determine endpoints local to a node.
+        NodeName *string  
+        // Optional: the conditions of the endpoint
+        Conditions EndpointConditions
+        // Optional: Reference to object providing the endpoint.
+        TargetRef *v1.ObjectReference
+}
+
+type EndpointConditions struct {
+       // Matches the Ready condition on pod
+       Ready           bool
+       // Matches ContainersReady condition on pod
+       ContainersReady bool
+}
+
+```
+
+### Mapping
```

Comment:
How are endpoints mapped to slices?

Reply:
It depends on the algorithm in controller below.